src/lib.rs

```diff
@@ -413,27 +388,29 @@ fn format_file(file: &Path, config: &Config) -> FileMap {
                                                 codemap.clone());
     let mut parse_session = ParseSess::with_span_handler(tty_handler, codemap.clone());
 
-    let krate = parse::parse_crate_from_file(file, Vec::new(), &parse_session).unwrap();
+    let main_file = match input {
+        Input::File(ref file) => file.clone(),
+        Input::Text(..) => PathBuf::from("stdin"),
+    };
+
+    let krate = match parse_input(input, &parse_session) {
+        Ok(krate) => krate,
+        Err(mut diagnostic) => {
+            diagnostic.emit();
+            panic!("Unrecoverable parse error");
+        }
+    };
```

Comment:
`unwrap` used to be here. However `DiagnosticBuilder` panics in the `drop`, if it wasn't `cancel`ed or `emit`ed, so this `unwrap` caused `panic wile panicking`. Now there is only one panic, which is an improvement I think :)